File: fn/src/main/java/org/sqlbuilder/fn/joins/FEPattern.java
package org.sqlbuilder.fn.joins;

import org.sqlbuilder.common.Insertable;
import org.sqlbuilder.common.RequiresIdFrom;
import org.sqlbuilder.fn.objects.FERealization;
import org.sqlbuilder.fn.objects.ValenceUnit;

import java.util.Comparator;
import java.util.HashSet;
import java.util.Set;

public class FEPattern extends Pair<FERealization, ValenceUnit> implements Insertable
{
	public static final Comparator<FEPattern> COMPARATOR = Comparator //
			.comparing(FEPattern::getLuId) //
			.thenComparing(FEPattern::getFeName) //
			.thenComparing(FEPattern::getSecond);

	public static final Set<FEPattern> SET = new HashSet<>();

	// C O N S T R U C T O R

	public static FEPattern make(final FERealization fer, final ValenceUnit vu)
	{
		var vr = new FEPattern(fer, vu);
		SET.add(vr);
		return vr;
	}

	private FEPattern(final FERealization fer, final ValenceUnit vu)
	{
		super(fer, vu);
	}

	// A C C E S S

	public FERealization getFer()
	{
		return first;
	}

	public String getFeName()
	{
		return first.getFEName();
	}

	public int getLuId()
	{
		return first.getLuId();
	}

	// I N S E R T

	@RequiresIdFrom(type = FERealization.class)
	@RequiresIdFrom(type = ValenceUnit.class)
	@Override
	public String dataRow()
	{
		return String.format("%s,%s", //
				first.getSqlId(), //
				second.getSqlId()); //
	}

	@Override
	public String comment()
	{
		return  String.format("luid=%d,fe=%s,vu={%s,%s,%s}", first.getLuId(), first.getFEName(), second.getFE(), second.getPT(), second.getGF());
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return String.format("[FER-VU fer=%s vu=%s]", first, second);
	}
}

File: fn/src/main/java/org/sqlbuilder/fn/joins/FEPattern.kt
package org.sqlbuilder.fn.joins

import org.sqlbuilder.annotations.RequiresIdFrom
import org.sqlbuilder.common.Insertable
import org.sqlbuilder.fn.objects.FERealization
import org.sqlbuilder.fn.objects.ValenceUnit
import java.util.Objects

data class FEPattern(
    val fer: FERealization,
    val vu: ValenceUnit,
) : Insertable {

    val feName: String
        get() = fer.fEName

    val luId: Int
        get() = fer.luId

    // I N S E R T

    @RequiresIdFrom(type = FERealization::class)
    @RequiresIdFrom(type = ValenceUnit::class)
    override fun dataRow(): String {
        return "${fer.getSqlId()},${vu.getSqlId()}"
    }

    override fun comment(): String {
        return "luid=${fer.luId},fe=${fer.fEName},vu={${vu.fE},${vu.pT},${vu.gF}}"
    }

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as FEPattern
        return fer == that.fer && vu == that.vu
    }

    override fun hashCode(): Int {
        return Objects.hash(fer, vu)
    }

    // T O S T R I N G

    override fun toString(): String {
        return "[FER-VU fer=$fer vu=$vu]"
    }

    companion object {

        val COMPARATOR: Comparator<FEPattern> = Comparator
            .comparing<FEPattern, Int> { it.luId }
            .thenComparing<String> { it.feName }
            .thenComparing<ValenceUnit> { it.vu }

        val SET = HashSet<FEPattern>()

        fun make(fer: FERealization, vu: ValenceUnit): FEPattern {
            val vr = FEPattern(fer, vu)
            SET.add(vr)
            return vr
        }
    }
}
